### PR TITLE
GasGeneratorComponent

### PR DESCRIPTION
**Pull request by collinlunn**
Generates gas in an attached pipe, for making a sustainable station atmospheric system.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+using Robust.Shared.GameObjects;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Log;
+using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+using System.Linq;
+
+namespace Content.Server.GameObjects.Components.Atmos.Piping
+{
+    [RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs:14]: Include your description as a doc comment here.

File: Resources/Textures/Constructible/Atmos/gasgenerator.rsi/meta.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+   "size":{
+      "x":32,
+      "y":32
+   },
+   "license":"CC-BY-SA-3.0",
+   "copyright":"Taken from https://github.com/tgstation/tgstation at commit 57cd1d59ca019dd0e7811ac451f295f818e573da",
+   "states":[
+      {
+         "name":"gasGenerator",
+         "directions":1
```
Comment (metalgearsloth) [inline Resources/Textures/Constructible/Atmos/gasgenerator.rsi/meta.json:12]: Nitpicky but directions not required anymore if it's 1.

```suggestion
         "name":"gasGenerator"
```

File: Resources/Prototypes/Entities/Constructible/Ground/gasgenerator.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  - type: InteractionOutline
+  - type: Physics
+  - type: SnapGrid
+    offset: Center
+  - type: GasGenerator
+    
+- type: entity
+  parent: GasGeneratorBase
+  id: GasGenerator
+  name: Gas Generator
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Constructible/Ground/gasgenerator.yml:17]:
```suggestion
  name: gas generator
```

File: Resources/Prototypes/Entities/Constructible/Ground/gasgenerator.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+- type: entity
+  abstract: true
+  id: GasGeneratorBase
+  placement:
+    mode: SnapgridCenter
+  components:
+  - type: Clickable
+  - type: InteractionOutline
+  - type: Physics
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Constructible/Ground/gasgenerator.yml:9]: Better to be explicit with physics but out of scope if you want.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Just minor things.

**Review by PJB3005 — APPROVED** [review]

File: Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                return;
+
+            Pipe.Air.AdjustMoles(GeneratedGas, GasGenerationRate);
+        }
+
+        private void SetPipes()
+        {
+            if (!Owner.TryGetComponent<NodeContainerComponent>(out var container))
+            {
+                Logger.Error($"{typeof(GasGeneratorComponent)} on {Owner?.Prototype?.ID}, Uid {Owner?.Uid} did not have a {nameof(NodeContainerComponent)}.");
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs:79]:
```suggestion
                Logger.Error($"{nameof(GasGeneratorComponent)} on {Owner?.Prototype?.ID}, Uid {Owner?.Uid} did not have a {nameof(NodeContainerComponent)}.");
```

File: Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+                return;
+            }
+            Pipe = container.Nodes.OfType<PipeNode>().FirstOrDefault();
+            if (Pipe == null)
+            {
+                Logger.Error($"{typeof(GasGeneratorComponent)} on {Owner?.Prototype?.ID}, Uid {Owner?.Uid} could not find compatible {nameof(PipeNode)}s on its {nameof(NodeContainerComponent)}.");
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Atmos/Piping/GasGeneratorComponent.cs:85]:
```suggestion
                Logger.Error($"{nameof(GasGeneratorComponent)} on {Owner?.Prototype?.ID}, Uid {Owner?.Uid} could not find compatible {nameof(PipeNode)}s on its {nameof(NodeContainerComponent)}.");
```

**Review by PaulRitter — CHANGES_REQUESTED** [review]
nitpicks

**Pull request merged**